org.opentosca.planbuilder.type.plugin.ubuntuvm/src/org/opentosca/planbuilder/type/plugin/ubuntuvm/AbstractPlugin.java

```diff
@@ -0,0 +1,119 @@
+package org.opentosca.planbuilder.type.plugin.ubuntuvm;
+
+import org.opentosca.container.core.tosca.convention.Types;
+import org.opentosca.container.core.tosca.convention.Utils;
+import org.opentosca.planbuilder.model.tosca.AbstractNodeTemplate;
+import org.opentosca.planbuilder.model.tosca.AbstractRelationshipTemplate;
+import org.opentosca.planbuilder.plugins.IPlanBuilderTypePlugin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractPlugin implements IPlanBuilderTypePlugin {
+	private static final String ID = "OpenTOSCA PlanBuilder VM and Cloud Provider Declarative Type Plugin";
+	private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPlugin.class);
+
+	@Override
+	public String getID() {
+		return ID;
+	}
+
+	@Override
+	public boolean canHandle(AbstractNodeTemplate nodeTemplate) {
+		if (nodeTemplate == null) {
+			LOGGER.debug("NodeTemplate is null");
+		}
+		if (nodeTemplate.getType() == null) {
+			LOGGER.debug("NodeTemplate NodeType is null. NodeTemplate Id:" + nodeTemplate.getId());
+		}
+		if (nodeTemplate.getType().getId() == null) {
+			LOGGER.debug("NodeTemplate NodeType id is null");
+		}
+		// this plugin can handle all referenced nodeTypes
+		if (Utils.isSupportedCloudProviderNodeType(nodeTemplate.getType().getId())) {
+			return true;
+		} else if (Utils.isSupportedVMNodeType(nodeTemplate.getType().getId())) {
+			// checking if this vmNode is connected to a nodeTemplate of Type
+			// cloud provider (ec2, openstack) or docker engine, if not this
+			// plugin can't handle
+			// this node
+			for (final AbstractRelationshipTemplate relationshipTemplate : nodeTemplate.getOutgoingRelations()) {
+				if (Utils.isSupportedCloudProviderNodeType(relationshipTemplate.getTarget().getType().getId())
+						| Utils.isSupportedDockerEngineNodeType(relationshipTemplate.getTarget().getType().getId())) {
+					return true;
+				}
+			}
+			return false;
+		} else if (Utils.isSupportedInfrastructureNodeType(nodeTemplate.getType().getId())) {
```

Comment:
else is also not necessary here, since all codepaths in the previous blocks result in a return